File: src/test/java/com/testainers/StatusResourceTest.java
package com.testainers;

import io.quarkus.test.junit.QuarkusTest;
import io.restassured.http.Method;
import org.junit.jupiter.api.Test;

import static org.hamcrest.Matchers.*;

/**
 * @author Eduardo Folly
 */
@QuarkusTest
public class StatusResourceTest extends BaseResourceTest {

    @Test
    public void statusString() {
        for (Method method : METHODS) {
            json(method)
                    .request(method, "/status/a")
                    .then()
                    .statusCode(404)
                    .statusLine(containsStringIgnoringCase("Not Found"));
        }
    }

    @Test
    public void statusDouble() {
        for (Method method : METHODS) {
            json(method)
                    .request(method, "/status/1.8")
                    .then()
                    .statusCode(404)
                    .statusLine(containsStringIgnoringCase("Not Found"));
        }
    }

    @Test
    public void statusNegative() {
        for (Method method : METHODS) {
            json(method).request(method, "/status/-1")
                        .then()
                        .statusCode(500)
                        .body("body", is("Unknown status code: -1"),
                              bodyMatchers(method));
        }
    }

    @Test
    public void status0() {
        for (Method method : METHODS) {
            json(method).request(method, "/status/0")
                        .then()
                        .statusCode(500)
                        .body("body", is("Unknown status code: 0"),
                              bodyMatchers(method));
        }
    }

    @Test
    public void status99() {
        for (Method method : METHODS) {
            json(method).request(method, "/status/99")
                        .then()
                        .statusCode(500)
                        .body("body", is("Unknown status code: 99"),
                              bodyMatchers(method));
        }
    }

    @Test
    public void status100() {
        for (Method method : METHODS) {
            json(method).request(method, "/status/100")
                        .then()
                        .statusCode(500)
                        .body("body",
                              is("Informational responses are not supported: 100"),
                              bodyMatchers(method));
        }
    }

    @Test
    public void status199() {
        for (Method method : METHODS) {
            json(method)
                    .request(method, "/status/199")
                    .then()
                    .statusCode(500)
                    .body("body",
                          is("Informational responses are not supported: 199"),
                          bodyMatchers(method));
        }
    }

    @Test
    public void status200() {
        for (Method method : METHODS) {
            json(method)
                    .request(method, "/status/200")
                    .then()
                    .statusCode(200)
                    .body("body",
                          method == Method.GET ? not(BODY) : is(BODY),
                          bodyMatchers(method));
        }
    }

    @Test
    public void status204() {
        for (Method method : METHODS) {
            json(method).request(method, "/status/204")
                        .then()
                        .statusCode(204)
                        .body(is(""));
        }
    }

    @Test
    public void status205() {
        for (Method method : METHODS) {
            json(method).request(method, "/status/205")
                        .then()
                        .statusCode(205)
                        .headers("content-length", "0")
                        .body(is(""));
        }
    }

    @Test
    public void status304() {
        for (Method method : METHODS) {
            json(method).request(method, "/status/304")
                        .then()
                        .statusCode(304)
                        .body(is(""));
        }
    }

    @Test
    public void status599() {
        for (Method method : METHODS) {
            json(method).request(method, "/status/599")
                        .then()
                        .statusCode(599)
                        .body("body",
                              method == Method.GET ? not(BODY) : is(BODY),
                              bodyMatchers(method));
        }
    }

    @Test
    public void status600() {
        for (Method method : METHODS) {
            json(method).request(method, "/status/600")
                        .then()
                        .statusCode(500)
                        .body("body", is("Unknown status code: 600"),
                              bodyMatchers(method));
        }
    }

}

File: src/test/kotlin/com/testainers/StatusResourceTest.kt
package com.testainers

import io.quarkus.test.junit.QuarkusTest
import io.restassured.http.Method
import org.hamcrest.Matchers.*
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource

/**
 * @author Eduardo Folly
 */
@QuarkusTest
class StatusResourceTest : BaseResourceTest() {
    companion object : BaseResourceTest() {
        @JvmStatic
        fun unknownStatus(): List<Arguments> =
            argumentGenerator(listOf(-1, 0, 1, 99, 600))

        @JvmStatic
        fun informationalStatus(): List<Arguments> =
            argumentGenerator(listOf(100, 199))

        @JvmStatic
        fun validStatus(): List<Arguments> = argumentGenerator(listOf(200, 599))

        @JvmStatic
        fun emptyStatus(): List<Arguments> =
            argumentGenerator(listOf(204, 205, 304))
    }

    @ParameterizedTest
    @MethodSource("notFoundStatus")
    fun notFound(
        method: Method,
        status: String?,
    ) {
        json(method)
            .request(method, "/status/$status")
            .then()
            .statusCode(404)
            .statusLine(containsStringIgnoringCase("Not Found"))
    }

    @ParameterizedTest
    @MethodSource("unknownStatus")
    fun unknown(
        method: Method,
        status: Int,
    ) {
        json(method)
            .request(method, "/status/$status")
            .then()
            .statusCode(500)
            .body(
                "body",
                equalTo("Unknown status code: $status"),
                *bodyMatchers(method),
            )
    }

    @ParameterizedTest
    @MethodSource("informationalStatus")
    fun informational(
        method: Method,
        status: Int,
    ) {
        json(method)
            .request(method, "/status/$status")
            .then()
            .statusCode(500)
            .body(
                "body",
                equalTo("Informational responses are not supported: $status"),
                *bodyMatchers(method),
            )
    }

    @ParameterizedTest
    @MethodSource("validStatus")
    fun valid(
        method: Method,
        status: Int,
    ) {
        json(method)
            .request(method, "/status/$status")
            .then()
            .statusCode(status)
            .body(
                "body",
                if (method == Method.GET) not(body) else equalTo(body),
                *bodyMatchers(method),
            )
    }

    @ParameterizedTest
    @MethodSource("emptyStatus")
    fun empty(
        method: Method,
        status: Int,
    ) {
        json(method)
            .request(method, "/status/$status")
            .then()
            .statusCode(status)
            .body(equalTo(""))
    }
}
